GCP AlloyDB support

Tener opened the pull request:
Closes https://github.com/gravitational/teleport/issues/23336.

changelog: Added support for GCP AlloyDB.

Tener committed:
Types changes

Tener committed:
Configuration changes

Tener committed:
Engine changes

Tener committed:
Proto changes

Tener committed:
Cloud client, `go.mod`

Tener committed:
Add method to fetch AlloyDB token

Tener committed:
`make -C integrations/operator crd`

Tener committed:
lint fix

Tener committed:
`make -C integrations/terraform docs`

greedy52 reviewed (commented):
LGTM. i just need to test this out next week

Tener commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -586,12 +586,18 @@ message DocumentDB {
   string ProjectID = 1 [(gogoproto.jsontag) = "project_id,omitempty"];
   // InstanceID is the Cloud SQL instance ID.
   string InstanceID = 2 [(gogoproto.jsontag) = "instance_id,omitempty"];
+  // Region is the GCP region. Required for AlloyDB databases.
+  string Region = 3 [(gogoproto.jsontag) = "region,omitempty"];
```
I went with region because Google uses that language commonly and I find it much easier to understand than "location".

greedy52 replied:
Thanks for confirming. We just need to make sure the documentation is clear on which one to input.

Tener commented on lib/srv/db/postgres/connector.go:
```diff
@@ -145,6 +148,69 @@ func (c *connector) getConnectConfig(ctx context.Context) (*pgconn.Config, error
+		// serverProxyPort is the port the server-side proxy receives connections on.
+		const serverProxyPort = 5433
+		if config.Port != serverProxyPort {
+			c.log.WarnContext(ctx, "Non-standard AlloyDB port configured", "expected", serverProxyPort, "actual", config.Port)
```
Do you mean by `tctl` or some other way?

In any case I'd keep this one here as this wrong port is highly unlikely to be correct and this location makes it easier to find in the logs when the likely follow-up failure happens.

Tener replied:
I'd rather avoid hard failures like this; if the port is incorrect then the exchange will simply fail. But I guess correct port is another thing to outline in bold in the docs.

Tener committed:
change GCPCloudSQL comment in proto

marcoandredinis commented on lib/srv/db/postgres/alloydb.go:
```diff
@@ -0,0 +1,92 @@
+
+	// check
+	if mdxResp.GetResponseCode() != connectorspb.MetadataExchangeResponse_OK {
+		return trace.BadParameter("metadata exchange failed: %v", mdxResp.GetResponseCode())
```
```suggestion
		return trace.BadParameter("metadata exchange failed: %v, error: %s", mdxResp.GetResponseCode(), mdxResp.GetError())
```

Tener committed:
Update lib/srv/db/postgres/alloydb.go

Co-authored-by: Marco Dinis <marco.dinis@goteleport.com>

marcoandredinis commented on api/types/database.go:
```diff
@@ -965,6 +977,29 @@ func (d *DatabaseV3) IsEqual(i Database) bool {
 }
 
+// handleAlloyDBConfig validates AlloyDB configuration
+func (d *DatabaseV3) handleAlloyDBConfig() error {
```
I would also validate the port in the URI.
It must be either 5433 or absent from the URI (which I'm not sure we allow)

marcoandredinis commented on lib/srv/db/common/auth.go:
```diff
@@ -483,6 +485,24 @@ func (a *dbAuth) GetCloudSQLAuthToken(ctx context.Context, databaseUser string)
+// GetAlloyDBAuthToken returns authorization token that will be used as a
+// password when connecting to AlloyDB databases.
+func (a *dbAuth) GetAlloyDBAuthToken(ctx context.Context, databaseUser string) (string, error) {
+	//   https://cloud.google.com/alloydb/docs/connect-iam#procedure
```
```suggestion
	// https://cloud.google.com/alloydb/docs/connect-iam#procedure
```

marcoandredinis commented on lib/cloud/gcp/alloydb.go:
```diff
@@ -0,0 +1,90 @@
+	clock  clockwork.Clock
+}
+
+func (g *gcpAlloyDBAdminClient) GenerateClientCertificate(ctx context.Context, db types.Database, certExpiry time.Time, pkey *keys.PrivateKey) (*tls.Certificate, string, error) {
```
Missing godoc
Some references to GCP docs/code would be helpful

marcoandredinis commented on lib/srv/db/postgres/connector.go:
```diff
@@ -145,6 +148,69 @@ func (c *connector) getConnectConfig(ctx context.Context) (*pgconn.Config, error
+			err = metadataExchangeAlloyDB(token, tlsConn)
+			if err != nil {
+				c.log.WarnContext(ctx, "Metadata exchange failed", "err", err)
+				_ = tlsConn.Close() // best effort
```
We could still log the error if it failed.

Tener committed:
Update lib/srv/db/common/auth.go

Co-authored-by: Marco Dinis <marco.dinis@goteleport.com>

Tener committed:
address review comments

Tener committed:
lint fix

greedy52 reviewed (commented):
Tested and worked as expected 🎉 . left some comments on configuration paramters and auth error

Tener commented on lib/config/fileconf.go:
```diff
@@ -2071,6 +2071,12 @@ type DatabaseGCP struct {
 	ProjectID string `yaml:"project_id,omitempty"`
 	// InstanceID is the Cloud SQL database instance ID.
 	InstanceID string `yaml:"instance_id,omitempty"`
+	// Region is the GCP region. Required for AlloyDB databases.
+	Region string `yaml:"region,omitempty"`
+	// ClusterID is the ID of the cluster. Required for AlloyDB databases.
+	ClusterID string `yaml:"cluster_id,omitempty"`
+	// IsAlloyDB is true if the database is an AlloyDB server.
+	IsAlloyDB bool `yaml:"is_alloy_db,omitempty"`
```
`is_alloydb` sounds good, I'll update.

Data modelling-wise I consider the type name alone `GCPCloudSQL` to be legacy. If we were to simply call it `GCP` it would be consistent with other clouds which simply use `AWS` or `Azure`.

Aside from the boolean, the new fields are not particularly unique to AlloyDB. I'd avoid putting them into a separate message for encapsulation. I think we are fine to continue with current approach, which is to simply ignore those fields if they are not relevant.

We could reshape the GCP section to be more like AWS one, but it would require substantial effort for little actual gain, if any.

greedy52 replied:
UX should be the priority after we can secure the connection. the need to configure FIVE fields in additional to an ip with possible chance using the wrong port isn't very good UX in my opinion.

i do think the best way to resolve this is auto-discovery which likely we won't be able to do in the near term. so anything that can improve the manual configuration process is welcome.

Tener replied:
Once you set the boolean we start checking those other fields too. We don't enforce port, but the warning will be sent. I can enforce the port, but that may need to be lifted in the future and cause issues at that point?

Tener commented on lib/srv/db/postgres/alloydb.go:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	// check
+	if mdxResp.GetResponseCode() != connectorspb.MetadataExchangeResponse_OK {
+		return trace.BadParameter("metadata exchange failed: %v, error: %s", mdxResp.GetResponseCode(), mdxResp.GetError())
+	}
```
Good call, I'll make the change.

Tener committed:
Rename `is_alloy_db` to `is_alloydb`

marcoandredinis reviewed (approved):
Changes look good 👍 
I've also tested the database access and I was able to access the database

Tener committed:
Clarify comments for AlloyDB database fields in DatabaseGCP struct

Tener committed:
Nudge user towards proper permissions

Tener commented on lib/srv/db/postgres/connector.go:
```diff
@@ -205,6 +206,22 @@ func (c *connector) getConnectConfig(ctx context.Context) (*pgconn.Config, error
+
+	 Make sure that AlloyDB user %q exists and has the following roles:
+	 - Cloud AlloyDB Database User
+	 - Cloud AlloyDB Client
```
As far as I can tell this is the minimal combo that works. I can try to boil it down to particular permissions, but my gut feeling is that `alloydb.instances.connect` is the missing part. `Cloud AlloyDB Client` is the only low-privilege role that provides it.

<img width="673" height="130" alt="image" src="https://github.com/user-attachments/assets/861c2b83-96e2-4539-8ee5-2e76a5aaf9d2" />

Tener commented on lib/srv/db/postgres/connector.go:
```diff
@@ -145,6 +149,84 @@ func (c *connector) getConnectConfig(ctx context.Context) (*pgconn.Config, error
+			}
+
+			tlsConn := tls.Client(conn, tlsConfig)
+			err = tlsConn.HandshakeContext(ctx)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			err = metadataExchangeAlloyDB(token, tlsConn)
+			if err != nil {
```
There are pros and cons to both styles and I'm not convinced the proposed code is better. I'd rather not discuss this level of stylistic choices.

Tener commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -586,12 +586,19 @@ message DocumentDB {
   string ProjectID = 1 [(gogoproto.jsontag) = "project_id,omitempty"];
   // InstanceID is the Cloud SQL instance ID.
   string InstanceID = 2 [(gogoproto.jsontag) = "instance_id,omitempty"];
+  // Region is the GCP region. Required only for AlloyDB databases.
+  string Region = 3 [(gogoproto.jsontag) = "region,omitempty"];
+  // ClusterID is the ID of the cluster. Required only for AlloyDB databases.
+  string ClusterID = 4 [(gogoproto.jsontag) = "cluster_id,omitempty"];
+  // IsAlloyDB is true if the database is an AlloyDB server.
+  bool IsAlloyDB = 5 [(gogoproto.jsontag) = "is_alloydb,omitempty"];
```
This was already discussed at length and rejected. Can't do without major refactor for no real gain.

Tener replied:
> just that all of the Alloy specific fields are encapsulated under the existing GCPCloudSQL message instead of what we have right now

So you mean to introduce another nested message just for AlloyDB?

greedy52 replied:
`instance_id` and `cluster_id` are specific to alloydb so i like them being under `gcp.alloydb`. OTOH `project_id`, and `region` are more general. they are like aws account id and region. though region is debatable as it is only required for some databases not the others. Like we might need to specify region for gcp redis as well.

how about:
```
  - name: "alloydb-primary"
    protocol: postgres
    uri: "11.22.33.44:5433"
    gcp:
      project_id: "teleport-dev-123456"
      region: "europe-west3"
      alloydb:
        cluster_id: "my-alloydb"
        instance_id: "alloydb-primary"
```

zmb3 replied:
Steve's suggestion looks fine to me.

The key take away is that UX should be the top priority for Q3. I will be happy to see features delayed because we spent extra time thinking about the best UX.

Tener replied:
PR updated for new configuration style. Examples:

1. Default configuration; endpoint not specified, default is `private` (VPC-only)

```yaml
  - name: "alloydb-instance"
    protocol: postgres
    uri: "alloydb://projects/my-project-123456/locations/europe-west1/clusters/my-cluster/instances/my-instance"
```

2. Endpoint override to `public` type (public IP is reachable from the wide internet)
```yaml
  - name: "alloydb-instance-public"
    protocol: postgres
    uri: "alloydb://projects/my-project-123456/locations/europe-west1/clusters/my-cluster/instances/my-instance"
    gcp:
      alloydb:
        endpoint: public
```

3. Endpoint override to particular IP address; bypasses API call for address resolution.
```yaml
  - name: "alloydb-instance-custom-ip"
    protocol: postgres
    uri: "alloydb://projects/my-project-123456/locations/europe-west1/clusters/my-cluster/instances/my-instance"
    gcp:
      alloydb:
        endpoint: "11.22.33.44"
```

Tener commented on lib/srv/db/postgres/alloydb.go:
```diff
@@ -0,0 +1,92 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if count != 4 {
```
This is just the number of bytes in 32 bit number we are reading. Not very magical but yeah, we can do it.

Tener committed:
Update lib/srv/db/postgres/alloydb.go

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

Tener committed:
remove todo in favor of longer comment

Tener commented on lib/srv/db/postgres/connector.go:
```diff
@@ -145,6 +149,84 @@ func (c *connector) getConnectConfig(ctx context.Context) (*pgconn.Config, error
+			ServerName:   config.Host,
+			MinVersion:   tls.VersionTLS13,
+			// TODO(Tener): copy other things from original TLS config?
+			// copy from original config
```
Kind of. The TODO was meant to indicate potential extension point. I've merged the TODO with regular comment to have one less TODO in the codebase. Only 2k more TODOs to go 😉

Tener committed:
reduce indentation

Tener committed:
suggest custom IAM role with minimal required scope

Tener committed:
Major implementation update:
- new URI format `alloydb://<connection uri>`
- removed some configuration fields; added `gcp.alloydb.endpoint` instead
- added connection URI resolution against configured endpoint type
- made query cancellation and health checks aware of AlloyDB needs
- unified query cancellation logic to use the `connector` type

Tener committed:
Revert description changes for flags; these are redundant for AlloyDB

greedy52 commented on tool/teleport/common/teleport.go:
```diff
@@ -259,8 +259,9 @@ func Run(options Options) (app *kingpin.Application, executedCommand string, con
-	dbStartCmd.Flag("gcp-instance-id", "(Only for Cloud SQL) GCP Cloud SQL instance identifier.").StringVar(&ccf.DatabaseGCPInstanceID)
+	dbStartCmd.Flag("gcp-project-id", "(Only for Cloud SQL) Project identifier.").StringVar(&ccf.DatabaseGCPProjectID)
+	dbStartCmd.Flag("gcp-instance-id", "(Only for Cloud SQL) Instance identifier.").StringVar(&ccf.DatabaseGCPInstanceID)
+	dbStartCmd.Flag("gcp-alloydb-endpoint", "(Only for AlloyDB) Endpoint type or IP address override.").StringVar(&ccf.DatabaseGCPAlloyDBEndpoint)
```
nit: lets list out the endpoint types

Tener commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -586,12 +586,28 @@ message DocumentDB {
+  // It supports two types of values:
+  // - An enum: one of "private", "public" or "psc". Corresponding type of address will be looked up using GCP API.
+  // - IP address. This address will be used directly, without querying the GCP API.
+  string Endpoint = 1 [(gogoproto.jsontag) = "endpoint,omitempty"];
```
Having the IP escape hatch can be useful in some advanced scenarios and troubleshooting. I'd rather keep it as is for additional control level it gives us.

greedy52 replied:
i agree it can be useful in certain situations but it's confusing to users in common situations what they need to set. we would like it as the "escape hatch", could we do a separate optional field?

Tener commented on api/types/database.go:
```diff
@@ -550,11 +561,20 @@ func (d *DatabaseV3) getGCPType() (string, bool) {
 	}
-	return "", false
+
+	if !gcp.AlloyDB.IsEmpty() {
+		return DatabaseTypeAlloyDB, true
+	}
```
Yes, we need it: if you forget to use `alloydb://` in URI (which is likely) but set `gcp.alloydb.endpoint`, then that is an easy way to tell the URI is misconfigured.

Tener commented on api/types/database.go:
```diff
@@ -965,6 +989,51 @@ func (d *DatabaseV3) IsEqual(i Database) bool {
+	}
+
+	// these fields can be derived from connection URI; make sure there is no discrepancy
+	if d.Spec.GCP.InstanceID == "" {
+		d.Spec.GCP.InstanceID = info.InstanceID
+	} else {
+		if d.Spec.GCP.InstanceID != info.InstanceID {
+			return trace.BadParameter("database %q GCP instance ID %q does not match the configured URI instance ID %q, "+
+				"omit the gcp.instance_id field and it will be derived automatically",
+				d.GetName(), d.Spec.GCP.InstanceID, info.InstanceID)
+		}
+	}
```
I've considered that path; in the end nothings stops anyone from manually setting `project_id`, so I think it makes sense to verify the consistency of this field.

The other option is to completely disallow `gcp.project_id` for AlloyDB. Thoughts?

Tener replied:
Ok; I'll change the validation logic to ensure these are both empty for AlloyDB instances.

Tener commented on lib/srv/db/endpoints/endpoints.go:
```diff
@@ -60,6 +60,8 @@ type ResolverBuilderConfig struct {
 	// GetGCPSQLAdminClient returns GCP Cloud SQL Admin client.
 	GetGCPSQLAdminClient(context.Context) (gcp.SQLAdminClient, error)
+	// GetGCPAlloyDBClient returns GCP AlloyDB Admin client.
+	GetGCPAlloyDBClient(context.Context) (gcp.AlloyDBAdminClient, error)
```
I think we should keep current code structure to keep it simple and easy to understand. This also makes it easier to tell which actions are executed using which account.

greedy52 commented on lib/srv/db/postgres/connector.go:
```diff
@@ -50,10 +59,65 @@ type connector struct {
+		var lastErr error
+
+		resolveFun := endpoints.ResolverFn(func(ctx context.Context) ([]string, error) {
+			sometimes.Do(func() {
```
i am not worried about changing ip. but more on when gcp api fails. if we are sticking to sometimes, i would use a shorter interval like 10minutes

Tener replied:
@GavinFrazar is it safe for `resolveFun` to block for up to several minutes? If we can safely make the retry internal then I'd try to do so.

greedy52 replied:
i imagine the case:
- the host SA is misconfigured
- health check gets a permission denied error
- user sees the error and fix the permission
- a minutes later health check turns green

GavinFrazar reviewed (approved):
Mostly just nitpick comments but LGTM

Tener committed:
Enhance error handling in ParseAlloyDBConnectionURI function and add comprehensive test cases for invalid URIs

Tener committed:
split `endpoint` into `endpoint_type` and `endpoint_override`

Tener committed:
Remove indentation from custom error messages

Tener committed:
Log resolved database address

Tener committed:
Merge branch 'master' into tener/alloydb-support

Tener committed:
fix serialization test

Tener committed:
explain the logic in `getGCPType`

Tener committed:
Merge branch 'master' into tener/alloydb-support

Tener commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -586,12 +586,38 @@ message DocumentDB {
+// AlloyDB contains AlloyDB specific configuration elements.
+message AlloyDB {
+  // EndpointType is the database endpoint type to use.
+  AlloyDBEndpointType EndpointType = 1 [(gogoproto.jsontag) = "endpoint_type,omitempty"];
```
Yeah, it was pretty painful. I'll switch back to string.

Tener committed:
use string instead of enum for endpoint type

Tener committed:
Merge branch 'master' into tener/alloydb-support

Tener committed:
add missing cast

Tener commented:
@r0mant @rosstimothy please approve

r0mant reviewed (approved):
@Tener Please don't forget to write an e2e guide as a follow up before this makes it into 18.2.
